pkg/policy/k8s/cell.go

```diff
@@ -65,69 +66,63 @@ type PolicyWatcherParams struct {
-	}
-
-	return &PolicyResourcesWatcher{
-		params: p,
+		return // skip watcher if K8s is not enabled
 	}
-}
-
-// WatchK8sPolicyResources starts watching Kubernetes policy resources.
-// Needs to be called before K8sWatcher.InitK8sSubsystem.
-func (p *PolicyResourcesWatcher) WatchK8sPolicyResources(ctx context.Context, policyManager PolicyManager) {
-	w := newPolicyWatcher(ctx, policyManager, p)
-	w.watchResources(ctx)
-}
 
-// newPolicyWatcher constructs a new policy watcher. Needs to be constructed
-// before K8sWatcher.InitK8sSubsystem is executed.
-// This constructor unfortunately cannot be started via the Hive lifecycle as
-// there exists a circular dependency between this watcher and the Daemon:
-// The constructor newDaemon cannot complete before all pre-existing
-// K8s policy resources have been added via the PolicyManager
-// (i.e. watchResources has observed the Sync event for CNP/CCNP/KNP).
-// Because the PolicyManager interface itself is implemented by the Daemon
-// struct, we have a circular dependency.
-func newPolicyWatcher(ctx context.Context, policyManager PolicyManager, p *PolicyResourcesWatcher) *policyWatcher {
-	// In order to not miss any service events, we register a subscriber here,
-	// before the call to K8sWatcher.InitK8sSubsystem.
-	svcCacheNotifications := stream.ToChannel(ctx, p.params.ServiceCache.Notifications(),
-		stream.WithBufferSize(int(p.params.Config.K8sServiceCacheSize)))
-
-	w := &policyWatcher{
-		log:                              p.params.Logger,
-		config:                           p.params.Config,
-		k8sResourceSynced:                p.params.K8sResourceSynced,
-		k8sAPIGroups:                     p.params.K8sAPIGroups,
-		svcCache:                         p.params.ServiceCache,
-		svcCacheNotifications:            svcCacheNotifications,
-		policyManager:                    policyManager,
-		ciliumNetworkPolicies:            p.params.CiliumNetworkPolicies,
-		ciliumClusterwideNetworkPolicies: p.params.CiliumClusterwideNetworkPolicies,
-		ciliumCIDRGroups:                 p.params.CiliumCIDRGroups,
-		networkPolicies:                  p.params.NetworkPolicies,
-
-		cnpCache:          make(map[resource.Key]*types.SlimCNP),
-		cidrGroupCache:    make(map[string]*cilium_v2_alpha1.CiliumCIDRGroup),
-		cidrGroupPolicies: make(map[resource.Key]struct{}),
-
-		toServicesPolicies: make(map[resource.Key]struct{}),
-		cnpByServiceID:     make(map[k8s.ServiceID]map[resource.Key]struct{}),
-	}
-	return w
+	// We want to subscribe before the start hook is invoked in order to not miss
+	// any events
+	ctx, cancel := context.WithCancel(context.Background())
+	svcCacheNotifications := stream.ToChannel(ctx, p.ServiceCache.Notifications(),
+		stream.WithBufferSize(int(p.Config.K8sServiceCacheSize)))
+
+	p.Lifecycle.Append(cell.Hook{
+		OnStart: func(startCtx cell.HookContext) error {
+			policyManager, err := p.PolicyManager.Await(startCtx)
+			if err != nil {
+				return err
+			}
```

Comment:
We can't revert this commit as is. `w.watchResources(ctx)` **needs** to be called before `InitK8sSubsystem`. If we wait for the Daemon promise here, then `watchResources` is called after `InitK8sSubsystem` and thus we're not waiting for policy ingestion to initiate endpoint regeneration.

I think we needd https://github.com/cilium/cilium/pull/32847 before we can revert this commit

Reply:
Good call! I already had done the `registerResourceWithSyncFn` for CiliumEnvoyConfig (3rd commit) in the constructor instead of the start hook, just for this reason.

Just force pushed new 2nd commit that adds that change to the revert, please take a look!